cpp/src/reductions/reductions.cu

```diff
@@ -39,234 +37,155 @@ void gpu_reduction_op(const T *data, const gdf_valid_type *mask,
-            F2 functor2;
-            Ld2 loader2;
-
-            launch_once(output, nullptr, output_size,
-                        output, 1, identity, functor2, loader2);
-            CUDA_CHECK_LAST();
-        }
-
-        return GDF_SUCCESS;
-    }
 
-    template <typename Functor, typename Loader>
-    static
-    void launch_once(const T *data, gdf_valid_type *valid, gdf_size_type size,
-                     T *output, gdf_size_type output_size, T identity,
-                     Functor functor, Loader loader) {
-        // find needed gridsize
-        // use atmost REDUCTION_BLOCK_SIZE blocks
-        int blocksize = REDUCTION_BLOCK_SIZE;
-        int gridsize = (output_size < REDUCTION_BLOCK_SIZE?
-                        output_size : REDUCTION_BLOCK_SIZE);
-
-        // launch kernel
-        gpu_reduction_op<<<gridsize, blocksize>>>(
-            // inputs
-            data, valid, size,
-            // output
-            output,
-            // action
-            functor,
-            // identity
-            identity,
-            // loader
-            loader
-        );
-    }
-
-};
-
-
-struct DeviceSum {
-    typedef IdentityLoader Loader;
-    typedef DeviceSum second;
-
-    template<typename T>
-    __device__
-    T operator() (const T &lhs, const T &rhs) {
-        return lhs + rhs;
-    }
-
-    template<typename T>
-    static constexpr T identity() { return T{0}; }
-};
-
-struct DeviceProduct {
-    typedef IdentityLoader Loader;
-    typedef DeviceProduct second;
-
-    template<typename T>
-    __device__
-    T operator() (const T &lhs, const T &rhs) {
-        return lhs * rhs;
-    }
-
-    template<typename T>
-    static constexpr T identity() { return T{1}; }
-};
-
-struct DeviceSumOfSquares {
-    struct Loader {
-        template<typename T>
-        __device__
-        T operator() (const T* ptr, int pos) const {
-            T val = ptr[pos];   // load
-            return val * val;   // squared
-        }
-    };
-    // round 2 just uses the basic sum reduction
-    typedef DeviceSum second;
-
-    template<typename T>
-    __device__
-    T operator() (const T &lhs, const T &rhs) const {
-        return lhs + rhs;
+    // First thread of each block stores the result.
+    if (tid == 0){
+        genericAtomicOperation(results[0], agg, functor);
     }
+}
 
-    template<typename T>
-    static constexpr T identity() { return T{0}; }
+template<typename T_in, typename T_out, typename Op>
+void ReduceOp(const gdf_column *input,
+                   gdf_scalar* scalar, cudaStream_t stream)
+{
+    T_out identity = Op::template identity<T_out>();
+
+    // allocate temporary memory for the result
+    void *result = NULL;
+    RMM_TRY(RMM_ALLOC(&result, sizeof(T_out), stream));
+
+    // initialize output by identity value
+    CUDA_TRY(cudaMemcpyAsync(result, &identity,
+            sizeof(T_out), cudaMemcpyHostToDevice, stream));
+    CUDA_CHECK_LAST();
+
+    int blocksize = reduction_block_size;
```

Comment:
Consider using the CUDA occupancy API to choose a good block / grid size...  Note your kernel has a grid-stride loop, so you don't necessarily need to launch as many threads as there are values as you do here.

Reply:
It seems that constant block size is still needed because `cub::BlockReduce` needs constant block size.

https://github.com/rapidsai/cudf/blob/9ae303077cd4a92264f743fcab2ae96478b11a3a/cpp/src/reductions/reductions.cu#L30